### DISCUSSION TITLE
Fetch historical data

### DISCUSSION THREAD

**Discussion by jb101010-2** [jb101010-2/pySuez#2] [Ideas]
Original discussion on https://github.com/ooii/pySuez/issues/28#issuecomment-2416408847
> I'm trying new improvments, I might have found a way to get and store the data directly at the right point of time, but not by using a sensor, I don't know if it's a real problem or not. By doing that I would be able to get all the data available from suez (even for new integrations) and save the data for the correct day and not the day suez provide it.
> 
> I have another question, on your screens you mentionned having green and red lines showing expected values, can you explain how they are computed ? directly by HA ? or something you have added on top of it ?

**Comment by jb101010-2** [comment]
Response from @kelesis

> Ok it makes sense to do it in 2 steps for quality reasons.
> 
> Wow, store data directly would be the best solution of course!
> 
> We could use built-in home assistant tools directly on history data
> There would be no time shift in the data indeed.
> But I wonder if there's a way to keep track track that storing history data has already been done on a first call to prevent doing this at each refresh ?
> It would be no problem for me to adapt my graphs based on home assistant history as it's a far better integrated solution.
> I can provide you or even publish all the yaml config, for my graphs, I use RomRider ApexCharts Card.
> For the lines, I compute them myself since ApexChart allows to generate data by using JavaScript directly inside yaml configuration.
> The code return an array of 2 values arrays [ [date, value], [date, value] ] with computed values.
> 
> ```
> data_generator: >
>       var year = new Date().getFullYear(); return [
> 
>       [new Date('0000-01-01'),null],
> 
>       [new Date((year-1)+'-01-01'),entity.attributes.last_year_overall/1000],
> 
>       [new Date((year+0)+'-01-01'),((new Date('12/31/'+year)-new
>       Date('1/1/'+year))/(new Date()-new
>       Date('1/1/'+year)))*entity.attributes.this_year_overall/1000]
> 
>       ];
> ```
> 
> The first 2 values are needed by apexchart for correct timescale on X coordinates. The second is past year values (in Liters / 1000 for m³). The last one is a simple cross multiplication ("règle de trois")
> to evaluate where I will be at the end of this year if consumption keeps at the same level it's been until now.
> 
> Same thing for the cost.
> 
> Not sure it is the right place to copy paste the whole yaml? 😅 Or maybe yes tell me.

**Reply by jb101010-2** [reply]
I should be able to fetch all available historical data on first call, and only new data on following calls.  
But if you what you mean is aggregate it to data you have already fetched I don't know if it would be possible.

Seeing how you generate future data, for now I don't think it should be provided along real data.
Maybe be available somewhere else, but I don't have an idea of a good place for now, I will keep it in mind if I find a use or a place for it.

I don't think I need your configuration for now, but if you think it can help you could attach it to a reply.

**Reply by jb101010-2** [reply]
I have successfuly retrieved the data and integrated in HA statistic cards.  
![image](https://github.com/user-attachments/assets/02c1108b-6df3-4a1b-8f37-a3ac6f14993f)

and in ha energy dahsboard
![image](https://github.com/user-attachments/assets/f53624c7-d048-4161-8f53-37be0cbc8387)

missing the cost in energy dashboard for now

**Reply by Kelesis** [reply]
"But if you what you mean is aggregate it to data you have already fetched I don't know if it would be possible."
No no just not making the call at each refresh is perfect 👌

**Reply by jb101010-2** [reply]
Ok so I think everything is working. 
I have added another statistics representing the price, unfortunately the historical price will be computed according to the current, so it could be wrong, but future data will be ok.

I have discovered one thing, the cost isn't displayed in energy dashboard graphs, only in the table below it, (or I haven't found out how) if not I'll try to get in touch with the HA team to see if they'd be interrested in an evolution of the cards.

Having all these statistical data I don't know if the sensor existing in the integration are useful anymore.
- total counter, useful only for energy dashboard
- old historical sensor no more usage
- maybe one with last known value
